### PR TITLE
added scaling-related changes

### PR DESCRIPTION
**Pull request by BatiGencho**
- fixed coins, fee and rolls scaling for obtaining correct amounts
- added conversion functions toMAS and fromMAS for ready use by other libraries
- revised test example
- removed sc-utils dependency
- replaced bn.js with much better bignumber.js library allowing decimals

### CODE REVIEW

File: examples/smartContracts/deployer.ts [hunk 1]
```diff
@@ -24,85 +57,75 @@ export const deploySmartContract = async (
...
-		}
-		if (!compiledSc.binary) {
-			const msg = chalk.red(`No bytecode to deploy for wasm file ${chalk.yellow(deploymentScWasm)}. Check AS compiler`);
-			console.error(msg);
-			throw new Error(msg);
-		}
+		// set the number of contracts
+		datastore.set(
+		  new Uint8Array([0x00]),
+		  new Uint8Array(new Args().addU64(BigInt(contractsToDeploy.length)).serialize()),
```
Comment (peterjah) [inline examples/smartContracts/deployer.ts:69]: u64ToBytes(BigInt(contractsToDeploy.length))

File: examples/smartContracts/deployer.ts [hunk 2]
```diff
@@ -24,85 +57,75 @@ export const deploySmartContract = async (
...
+				contract.data,
+			);
+			if (contract.args) {
+				datastore.set(
+					new Uint8Array(
+					new Args()
+						.addU64(BigInt(i + 1))
+						.addUint8Array(new Uint8Array([0x00]))
+						.serialize(),
+					),
```
Comment (peterjah) [inline examples/smartContracts/deployer.ts:86]: new Uint8Array([...u64ToBytes(BigInt(i + 1)), 0])

File: examples/smartContracts/deployer.ts [hunk 3]
```diff
@@ -24,85 +57,75 @@ export const deploySmartContract = async (
...
-			const msg = chalk.red(`Error compiling smart contract on path ${chalk.yellow(deployedScWasm)}`);
-			console.error(msg);
-			throw new Error(ex);
-		}
-		if (!compiledDeployedSc.binary) {
-			const msg = chalk.red(`No bytecode to deploy for wasm file ${chalk.yellow(deployedScWasm)}. Check AS compiler`);
-			console.error(msg);
-			throw new Error(msg);
+			datastore.set(
+				new Uint8Array(new Args().addU64(BigInt(i + 1)).serialize()),
```
Comment (peterjah) [inline examples/smartContracts/deployer.ts:76]: u64ToBytes(BigInt(i + 1))

File: src/web3/BaseClient.ts [hunk 4]
```diff
@@ -198,7 +190,7 @@ export class BaseClient {
 				const maxGasEncoded = Buffer.from(varintEncode((data as ICallData).maxGas));
 
 				// coins to send
-				const coinsEncoded = Buffer.from(varintEncode((data as ICallData).coins));
+				const coinsEncoded = Buffer.from(varintEncode(toMAS((data as ICallData).coins)));
```
Comment (AurelienFT) [inline src/web3/BaseClient.ts:193]: This breaks a lot of things especially in the deployer we need to be careful when making the update maybe it should be a major.

**Review by AurelienFT — APPROVED** [review]

File: examples/smartContracts/deployer.ts [hunk 5]
```diff
@@ -24,85 +57,75 @@ export const deploySmartContract = async (
...
+				);
+			}
+			if (contract.coins > 0) {
+				datastore.set(
+					new Uint8Array(
+					new Args()
+						.addU64(BigInt(i + 1))
+						.addUint8Array(new Uint8Array([0x01]))
+						.serialize(),
+					),
```
Comment (peterjah) [inline examples/smartContracts/deployer.ts:97]: new Uint8Array([...u64ToBytes(BigInt(i + 1)), 1])

**Comment by AurelienFT** [comment]
I add @Thykof so that he can be aware of the changes on this PR.

File: examples/smartContracts/deployer.ts [hunk 6]
```diff
@@ -24,85 +57,75 @@ export const deploySmartContract = async (
...
+					new Uint8Array(new Args().addU64(BigInt(contract.coins)).serialize()),
```
Comment (peterjah) [inline examples/smartContracts/deployer.ts:98]: u64ToBytes(BigInt(contract.coins))

File: src/utils/converters.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+    const amountScaled = amount.times(scaleFactor);
+    return parseInt(amountScaled.toFixed(0));
+};
+
+export const fromMAS = (inputAmount: number | string): BigNumber => {
+    const amount = new BigNumber(inputAmount);
+    const scaleFactor = (new BigNumber(10)).pow(new BigNumber(MASSA_SCALING_FACTOR));
+    const amountScaled = amount.dividedBy(scaleFactor);
+    return amountScaled;
+};
```
Comment (peterjah) [inline src/utils/converters.ts:17]: It would be nice to have unit tests with hedge cases for theses functions

File: src/utils/converters.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+    const scaleFactor = (new BigNumber(10)).pow(new BigNumber(MASSA_SCALING_FACTOR));
+    const amountScaled = amount.times(scaleFactor);
+    return parseInt(amountScaled.toFixed(0));
+};
+
+export const fromMAS = (inputAmount: number | string): BigNumber => {
+    const amount = new BigNumber(inputAmount);
+    const scaleFactor = (new BigNumber(10)).pow(new BigNumber(MASSA_SCALING_FACTOR));
+    const amountScaled = amount.dividedBy(scaleFactor);
+    return amountScaled;
```
Comment (peterjah) [inline src/utils/converters.ts:16]: maybe something like this is more efficient ?
   ```
    const amount = new BigNumber(inputAmount).toFixed(0);
    return BigNumber(amount.substring(0, amount.length - MASSA_SCALING_FACTOR));

**Review by peterjah — CHANGES_REQUESTED** [review]

File: examples/smartContracts/contracts/sc_code.ts [hunk 9]
```diff
@@ -1,6 +1,21 @@
...
+ *
+ * @param args - Arguments serialized with Args
+ */
+export function constructor(args: StaticArray<u8>): StaticArray<u8> {
+  // This line is important. It ensure that this function can't be called in the future.
+  // If you remove this check someone could call your constructor function and reset your SC.
+  if (!callerHasWriteAccess()) {
+    return [];
+  }
+  generateEvent(`Constructor called on contract ${Context.callee().toByteString()}`);
```
[the remark this answers is not in the document]
Comment (Thykof) [reply examples/smartContracts/contracts/sc_code.ts:15]: please use Address.toString() instead of Address.toByteString()

[the remark this answers is not in the document]
Comment (BatiGencho) [reply examples/smartContracts/contracts/sc_code.ts]: fixed

File: examples/smartContracts/contracts/sc_code.ts [hunk 10]
```diff
@@ -1,6 +1,21 @@
-import { print, generateEvent, call, Context, Address, toBytes, fromBytes, Storage } from "@massalabs/massa-as-sdk";
+import { print, generateEvent, call, Context, Address, toBytes, fromBytes, Storage, callerHasWriteAccess } from "@massalabs/massa-as-sdk";
```
[the remark this answers is not in the document]
Comment (BatiGencho) [reply examples/smartContracts/contracts/sc_code.ts]: fixed

File: examples/smartContracts/deployer.ts [hunk 11]
```diff
@@ -1,19 +1,52 @@
...
+interface ISCData {
+	data: Uint8Array;
+	args?: Args;
+	coins: number;
+}
+
+async function checkBalance(web3Client: Client, account: IAccount) {
+	const balance = await web3Client.wallet().getAccountBalance(account.address as string);
+	console.log("Wallet balance: ", balance?.final);
+	if (!balance?.final || !parseFloat(balance.final) || parseFloat(balance.final) === 0.0) {
```
[the remark this answers is not in the document]
Comment (BatiGencho) [reply examples/smartContracts/deployer.ts]: I have reworked that in a different way now

File: src/utils/converters.ts (hunk 7 above)
Comment (BatiGencho) [reply src/utils/converters.ts]: Yes, I will add some.

File: src/utils/converters.ts (hunk 8 above)
Comment (BatiGencho) [reply src/utils/converters.ts]: I have fixed that part now.

File: src/web3/MassaCoin.ts [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+import BigNumber from "bignumber.js";
+
+export class MassaCoin {
+    // unscaled original value
+    private value: BigNumber;
+
+    constructor(value: BigNumber | number | string) {
+        this.value = new BigNumber(value);
+        this.toValue = this.toValue.bind(this);
+        this.rawValue = this.rawValue.bind(this);
```
[the remark this answers is not in the document]
Comment (BatiGencho) [reply src/web3/MassaCoin.ts]: removed

**Review by peterjah — APPROVED** [review]

File: examples/smartContracts/deployer.ts (hunk 6 above)
Comment (BatiGencho) [reply examples/smartContracts/deployer.ts]: discussed and fixed

File: examples/smartContracts/deployer.ts (hunk 3 above)
Comment (BatiGencho) [reply examples/smartContracts/deployer.ts]: discussed and fixed

File: examples/smartContracts/deployer.ts (hunk 2 above)
Comment (BatiGencho) [reply examples/smartContracts/deployer.ts]: discussed and fixed

File: examples/smartContracts/deployer.ts (hunk 1 above)
Comment (BatiGencho) [reply examples/smartContracts/deployer.ts]: discussed and fixed

File: examples/smartContracts/deployer.ts (hunk 5 above)
Comment (BatiGencho) [reply examples/smartContracts/deployer.ts]: discussed and fixed
